token/sdk/vault/rws/rws.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package rws
+
+import (
+	"sync"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/fabric"
+	"github.com/hyperledger-labs/fabric-smart-client/platform/orion"
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view"
+	"github.com/hyperledger-labs/fabric-token-sdk/token"
+	fabric2 "github.com/hyperledger-labs/fabric-token-sdk/token/services/network/fabric"
+	orion2 "github.com/hyperledger-labs/fabric-token-sdk/token/services/network/orion"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/network/processor/rws"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/vault"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/vault/certification"
+	rws2 "github.com/hyperledger-labs/fabric-token-sdk/token/services/vault/rws"
+	"github.com/pkg/errors"
+)
+
+type VaultProvider struct {
+	sp view.ServiceProvider
+
+	vaultCacheLock sync.RWMutex
+	vaultCache     map[string]vault.TokenVault
+}
+
+func NewVaultProvider(sp view.ServiceProvider) *VaultProvider {
+	return &VaultProvider{sp: sp, vaultCache: make(map[string]vault.TokenVault)}
+}
+
+func (v *VaultProvider) Vault(network string, channel string, namespace string) (vault.TokenVault, error) {
+	k := network + channel + namespace
+	// Check cache
+	v.vaultCacheLock.RLock()
+	res, ok := v.vaultCache[k]
+	v.vaultCacheLock.RUnlock()
+	if ok {
+		return res, nil
+	}
+
+	// lock
+	v.vaultCacheLock.Lock()
+	defer v.vaultCacheLock.Unlock()
+
+	// check cache again
+	res, ok = v.vaultCache[k]
+	if ok {
+		return res, nil
+	}
+
+	tokenStore, err := rws.NewTokenStore(v.sp, token.TMSID{
+		Network:   network,
+		Channel:   channel,
+		Namespace: namespace,
+	})
+	if err != nil {
+		return nil, errors.WithMessagef(err, "failed to get token store")
+	}
+
+	storageProvider, err := certification.GetStorageProvider(v.sp)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get storage provider")
+	}
+
+	// Create new vault
+	if fns := fabric.GetFabricNetworkService(v.sp, network); fns != nil {
```

Comment:
This couples the Vault implementation with only two possible Networks (Fabric and Orion), but the code is almost the same between the two. Could it be an idea to hide the implementation behind an interface, so that as few components as possible care about the underlying network implementation?

Reply:
indeed, the responsibility of the provider is to return an instance of the interface vault.TokenVault.
One can change this provider in the sdk.go for full customisation. 
So, we need to care about the underlying backend only there. 
Anyhow, let's refine even more this part :)